libsolidity/analysis/ImmutableValidator.cpp

```diff
@@ -104,17 +105,47 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
 
 bool ImmutableValidator::visit(IfStatement const& _ifStatement)
 {
+	// store old context variables
 	bool prevInBranch = m_inBranch;
+	VariableDeclarationSet prevWrites = std::move(m_immutableStateVariablesWrittenInBranch);
```

Comment:
At this point, `m_immutableStateVariablesWrittenInBranch` is in an undefined state. We should clear it an assert it is cleared after the `accept`. Why is it safe to move it here anyway? Can't there be a write operation inside the if condition?

Reply:
That's not strictly true. It is in an `unspecified` state. 

from https://stackoverflow.com/questions/7027523/what-can-i-do-with-a-moved-from-object:

> Objects of types defined in the C++ standard library may be moved from (12.8). Move operations may be explicitly specified or implicitly generated. Unless otherwise specified, such moved-from objects shall be placed in a valid but unspecified state.

The main question now is if `emplace` is an operation that requires preconditions or not. I am still researching that question

Reply:
Or another phrasing from cppreference.com:

> in a "valid but unspecified state", meaning the object's class invariants hold (so functions without preconditions, such as the assignment operator, can be safely used on the object after it was moved from):

I interpret that, that it is completely okay to use `emplace` or `insert` after it was moved. I am not aware of any `valid` state that would prevent insertions in a `std::set` (e.g. you can't do anything to an instance to prevent `emplace` from working except memory exhaustion)

Reply:
That said, having an explicit clear after the move would def. make it failsafe and will avoid the next code reader wondering the same thing.

Reply:
Added explicit `clear` for applicable cases.